### PR TITLE
nekoray: init at 3.26

### LINKED ISSUE #244451: Package request: nekoray
**Project description**
Qt based cross-platform GUI proxy configuration manager (backend: v2ray / sing-box)

proxy configuration manager
**Metadata**

* homepage URL: https://github.com/MatsuriDayo/nekoray
* source URL: https://github.com/MatsuriDayo/nekoray
* license: gpl3
* platforms: unix, linux

### PR DESCRIPTION
**Pull request by TomaSajt**
## Description of changes

Closes https://github.com/NixOS/nixpkgs/issues/244451

I successfully built the `nekoray` package, and using 2 extra derivations with `buildGoModule`, I added the required `*_core` files as well.

The fetching is a bit weird, because I didn't want to deal with upstream's WEIRD way of fetching other repositories.

This needs some testing as I haven't got any experience with this program.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by TomaSajt** [comment]
Updated the PR to include the geodata files too.

**Comment by jys1670** [comment]
Result of `nixpkgs-review pr 300794` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>nekoray</li>
  </ul>
</details>

Tested `vless-xtls-reality`, works as expected. The same applies to `geosites.db` (didn't try other `.db` files, but they should be fine too).
The only broken thing I found is "Scan QR Code", sadly it crashes with exception:
```
terminate called after throwing an instance of 'std::invalid_argument'
  what():  Invalid image format
```
Earlier I was using a patched nekoray binary as a workaround, it has similar problem now, but used to work before (around December 2023). Looks like some dependencies were updated, causing incompatibilities. But it is a minor feature, nothing critical.

Thank you for packaging this one!

File: pkgs/by-name/ne/nekoray/package.nix [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+      hash = "sha256-6dlWDzI9ox4PQzEtJNgwA0pXmPC7fGrGId88Zl+1gpw=";
+    };
+  };
+
+  geodata = {
+    "geoip.dat" = fetchurl {
+      url = "https://github.com/Loyalsoldier/v2ray-rules-dat/releases/download/202404012210/geoip.dat";
+      hash = "sha256-0SOSwfSzee7UxJaa3HHgo6cUouRzGwUK17t9Xk+CLWw=";
+    };
+    "geosite.dat" = fetchurl {
```
Comment (Aleksanaa) [inline pkgs/by-name/ne/nekoray/package.nix:64]: This is packaged as https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/data/misc/v2ray-domain-list-community/default.nix

File: pkgs/by-name/ne/nekoray/package.nix [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    };
+    libneko = fetchSource {
+      name = "libneko";
+      rev = "5277a5bfc889ee7a89462695b0e678c1bd4909b1";
+      hash = "sha256-6dlWDzI9ox4PQzEtJNgwA0pXmPC7fGrGId88Zl+1gpw=";
+    };
+  };
+
+  geodata = {
+    "geoip.dat" = fetchurl {
```
Comment (Aleksanaa) [inline pkgs/by-name/ne/nekoray/package.nix:60]: This is packaged as https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/data/misc/v2ray-geoip/default.nix

File: pkgs/by-name/ne/nekoray/package.nix [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+      url = "https://github.com/Loyalsoldier/v2ray-rules-dat/releases/download/202404012210/geoip.dat";
+      hash = "sha256-0SOSwfSzee7UxJaa3HHgo6cUouRzGwUK17t9Xk+CLWw=";
+    };
+    "geosite.dat" = fetchurl {
+      url = "https://github.com/v2fly/domain-list-community/releases/download/20240331083526/dlc.dat";
+      hash = "sha256-VERU038ETTDSH3cehI7n2VbMOyqZ9+BAQ022LXKI0y4=";
+    };
+    "geoip.db" = fetchurl {
```
Comment (Aleksanaa) [inline pkgs/by-name/ne/nekoray/package.nix:68]: This is packaged as https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/data/misc/sing-geoip/default.nix

File: pkgs/by-name/ne/nekoray/package.nix [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+      url = "https://github.com/SagerNet/sing-geoip/releases/download/20240312/geoip.db";
+      hash = "sha256-d4eaQjntKKmDXn2x1/Vd0/lUCQRxQSXLOcw0JeYbf9Q=";
+    };
+    "geosite.db" = fetchurl {
```
Comment (Aleksanaa) [inline pkgs/by-name/ne/nekoray/package.nix:72]: I'd also recommend package this as well

File: pkgs/by-name/ne/nekoray/package.nix [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+  };
+
+  geodata = {
+    "geoip.dat" = "${v2ray-geoip}/share/v2ray/geoip.dat";
+    "geosite.dat" = "${v2ray-domain-list-community}/share/v2ray/geosite.dat";
+    "geoip.db" = "${sing-geoip}/share/sing-box/geoip.db";
+    "geosite.db" = fetchurl {
+      url = "https://github.com/SagerNet/sing-geosite/releases/download/20240331083526/geosite.db";
+      hash = "sha256-LLLdfOsLN3A69aXEXzSTH8Shc+pWtarHT2qy6zifehk=";
+    };
```
Comment (linsui) [inline pkgs/by-name/ne/nekoray/package.nix:70]: Isn't this already packaged?

```suggestion
    "geosite.db" = "${sing-geosite}/share/sing-box/geosite.db";
```

Comment (TomaSajt) [reply pkgs/by-name/ne/nekoray/package.nix]: Ah, thanks, forgot to check it.

**Comment by Aleksanaa** [comment]
Thanks for your contribution, I'll test it tomorrow. Also perhaps we need some capability wrappers or services for this, which can be done in another PR.

**Comment by TomaSajt** [comment]
Would anyone here like to co-maintain this package? I'm not actually someone who uses this program, I just like packaging stuff, so I don't really know how to use the program.

Because of this I'm probably not qualified to create a service for this ATM.

**Review by Aleksanaa — APPROVED** [review]
Tested basic functionality. Looks good.

**Comment by Aleksanaa** [comment]
@TomaSajt I will maintain the functionality of this package in the future. I will add myself to the maintainer in the next update.

**Comment by Aleksanaa** [comment]
@ofborg eval

**Comment by Aleksanaa** [comment]
@ofBorg build

File: pkgs/by-name/ne/nekoray/package.nix [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+  };
+
+  meta = {
+    description = "Qt based cross-platform GUI proxy configuration manager";
+    homepage = "https://github.com/MatsuriDayo/nekoray";
+    license = lib.licenses.gpl3Plus;
+    mainProgram = "nekoray";
+    maintainers = with lib.maintainers; [ tomasajt ];
+    platforms = lib.platforms.all;
```
Comment (Aleksanaa) [inline pkgs/by-name/ne/nekoray/package.nix:157]:
```suggestion
    platforms = lib.platforms.linux;
```

Builds fail on ofBorg, upstream also doesn't support Darwin very well

**Pull request merged**